### PR TITLE
Add validators

### LINKED ISSUE #9: Add validator to ProjectCreate, to check that project_dir is not an empty string
Refs:
* https://github.com/fractal-analytics-platform/fractal-web/issues/13
* https://github.com/fractal-analytics-platform/fractal-server/issues/525

### LINKED ISSUE #13: Add validators to `Workflow`-related schemas
1. In both `WorkflowCreate` and `WorkflowTaskCreate`, we could check that the fields representing IDs of other objects are non-negative.
2. In `WorkflowUpdate` we could check that `name` is not an empty string.

### LINKED ISSUE #14: Add validator to `UserCreate`
In `UserCreate`, we could check that `slurm_user` (if set) is not an empty string.

### LINKED ISSUE #15: Add validators to `Task`-related schemas
1. In `TaskUpdate`, all string attributes that are set should be non-empty.
2. The same holds for `TaskCreate`

### PR DESCRIPTION
**Pull request by ychiucco**
closes #9
closes #13 
closes #14 
closes #15 

ref #18

### CODE REVIEW

File: schemas/task.py [hunk 1]
```diff
@@ -46,19 +48,26 @@ class _TaskBase(SQLModel):
             Dictionary (saved as JSON) of the default parameters of the task
     """
 
-    name: str
-    source: str
-    
-    
+    name: str = Field(alias="name")
```
Comment (tcompa) [inline schemas/task.py:51]: These are not needed any more, right?

File: schemas/task.py [hunk 2]
```diff
@@ -46,19 +48,26 @@ class _TaskBase(SQLModel):
...
+    command: Optional[str]
+    source: Optional[str]
     default_args: Optional[Dict[str, Any]]  # type:ignore
     meta: Optional[Dict[str, Any]]  # type:ignore
 
+    @root_validator(pre=True)
+    def not_empty_strings(cls, values):
+        for k, v in values.items():
+            if isinstance(v, str):
+                values[k] = validate_str(v, k)
```
Comment (tcompa) [inline schemas/task.py:68]: This is not a validation check, but an actual change to the attribute.
For the moment I would rather stick with operations that do not modify the attribute, but rather "only" fail if it is not valid.

If we decide we want to be "smart" and modify the user's input, this should be a general decision - and we should probably do it in several places. At the moment, I propose we don't follow that path.

File: schemas/task.py [hunk 3]
```diff
@@ -84,6 +93,13 @@ class TaskCreate(_TaskBase):
     default_args: Optional[Dict[str, Any]] = Field(default={})
     meta: Optional[Dict[str, Any]] = Field(default={})
 
+    @root_validator(pre=True)
+    def not_empty_strings(cls, values):
+        for k, v in values.items():
+            if isinstance(v, str):
+                values[k] = validate_str(v, k)
```
Comment (tcompa) [inline schemas/task.py:100]: (same as above)

File: schemas/user.py [hunk 4]
```diff
@@ -20,3 +23,8 @@ class UserUpdate(schemas.BaseUserUpdate):
 
 class UserCreate(schemas.BaseUserCreate):
     slurm_user: Optional[str]
+
+    @validator("slurm_user")
+    def not_empty_str(cls, value):
+        v = validate_str(value, "slurm_user")
```
Comment (tcompa) [inline schemas/user.py:29]: (same as above, regarding modifications to the value)

File: schemas/workflow.py [hunk 5]
```diff
@@ -35,6 +36,12 @@ class WorkflowTaskCreate(_WorkflowTaskBase):
     task_id: int
     workflow_id: Optional[int]
 
+    @validator("task_id", "workflow_id")
+    def positive(cls, _id):
+        if (_id is not None) and _id < 1:
```
Comment (tcompa) [inline schemas/workflow.py:41]: Dumb question: do IDs always start at 1?

File: schemas/workflow.py [hunk 6]
```diff
@@ -74,10 +81,21 @@ class WorkflowRead(_WorkflowBase):
...
+            raise ValueError(f"IDs must be positive integers, given {_id}")
+        return _id
+
 
 class WorkflowUpdate(_WorkflowBase):
     name: Optional[str]  # type: ignore
 
+    @validator("name")
+    def not_empty_str(cls, value):
+        v = validate_str(value, "name")
```
Comment (tcompa) [inline schemas/workflow.py:96]: (same as above, regarding unexpected changes to attributes)

File: schemas/project.py [hunk 7]
```diff
@@ -36,11 +39,14 @@ class _ProjectBase(SQLModel):
 class ProjectCreate(_ProjectBase):
     default_dataset_name: Optional[str] = "default"
 
+    @validator("project_dir")
+    def not_empty_str(cls, value):
+        v = validate_str(value, "project_dir")
```
Comment (tcompa) [inline schemas/project.py:44]: (same as below, regarding unexpected changes to attributes)

File: schemas/task.py (hunk 2 above)
Comment (ychiucco) [reply schemas/task.py:68]: What should we do if the user enters this input `name="   name            "`?
Raise a `ValueError` or set `name="name"`?

File: schemas/__validator.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+def validate_str(string, attrname):
+    s = string.strip()
```
Comment (tcompa) [inline schemas/__validator.py:2]: Docstring

File: schemas/task.py (hunk 2 above)
Comment (tcompa) [reply schemas/task.py:68]: As per our last discussion: OK, let's strip the starting/ending whitespaces

File: schemas/workflow.py (hunk 5 above)
Comment (tcompa) [reply schemas/workflow.py:41]: @mfranzon ?

Comment (mfranzon) [reply schemas/workflow.py:41]: by default sql AUTOINCREMENT implementation starts from 1, so I think we are safe

**Pull request merged**